tests/cupy_tests/linalg_tests/test_decomposition.py

```diff
@@ -106,21 +106,56 @@ class TestSVD(unittest.TestCase):
     def setUp(self):
         self.seed = testing.generate_seed()
 
-    @testing.for_float_dtypes(no_float16=True)
+    @testing.for_dtypes([
+        numpy.int32, numpy.int64, numpy.uint32, numpy.uint64,
+        numpy.float32, numpy.float64, numpy.complex64, numpy.complex128,
+    ])
     def check_usv(self, shape, dtype):
         array = testing.shaped_random(
             shape, numpy, dtype=dtype, seed=self.seed)
         a_cpu = numpy.asarray(array, dtype=dtype)
         a_gpu = cupy.asarray(array, dtype=dtype)
         result_cpu = numpy.linalg.svd(a_cpu, full_matrices=self.full_matrices)
         result_gpu = cupy.linalg.svd(a_gpu, full_matrices=self.full_matrices)
-        self.assertEqual(len(result_cpu), len(result_gpu))
-        for b_cpu, b_gpu in zip(result_cpu, result_gpu):
-            # Use abs to support an inverse vector
-            cupy.testing.assert_allclose(
-                numpy.abs(b_cpu), cupy.abs(b_gpu), atol=1e-4)
+        # Check if the input matrix is not broken
+        cupy.testing.assert_allclose(a_gpu, a_cpu)
+
+        self.assertEqual(len(result_gpu), len(result_cpu))
+        u_cpu, s_cpu, vh_cpu = result_cpu
+        u_gpu, s_gpu, vh_gpu = result_gpu
+        cupy.testing.assert_allclose(s_gpu, s_cpu, atol=1e-4)
+
+        # assert unitary
+        cupy.testing.assert_allclose(
+            cupy.matmul(u_gpu.T.conj(), u_gpu),
+            numpy.eye(u_gpu.shape[1]),
+            atol=1e-4)
+        cupy.testing.assert_allclose(
+            cupy.matmul(vh_gpu, vh_gpu.T.conj()),
+            numpy.eye(vh_gpu.shape[0]),
+            atol=1e-4)
+
+        k, = s_cpu.shape
+        for j in range(k):
+            # assert corresponding vectors are equal up to rotation (`sign`)
+            uj_cpu = u_cpu[:, j]
+            vj_cpu = vh_cpu[j, :].conj()
+            uj_gpu = cupy.asnumpy(u_gpu[:, j])
+            vj_gpu = cupy.asnumpy(vh_gpu[j, :]).conj()
+            # Use least-squares estimation to compute rotation from cpu result
+            # to gpu result. We know norms of uj_cpu, vj_cpu are 1.
+            u_sign = numpy.vdot(uj_cpu, uj_gpu)
+            v_sign = numpy.vdot(vj_cpu, vj_gpu)
+            numpy.testing.assert_allclose(u_sign, v_sign, atol=1e-4)
+            sign = numpy.mean([u_sign, v_sign])
```

Comment:
Could you leave a comment on this line? I'm not sure I've understood it. Also wondering if it'd be bad checking the columns of U and V with their separate signs `u_sign`, `v_sign`.

Reply:
- `assert_allclose(u_sign, v_sign)` is necessary because the test does not compute `u @ diag(s) @ vh`.
- It looks good to separate other tests of `u` and `v`.